### PR TITLE
Guide for secondary entries.

### LINKED ISSUE #2764: Readme for creating secondary entries
- [x] I have written a [good issue](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Issue) 

## Describe the housekeeping request
Write a how-to guide on moving modules to secondary entries.

## Have you considered any alternatives?
No

## Are there any additional context?
No

### PR DESCRIPTION
**Pull request by kodeaben**
## Which issue does this PR close?

This PR closes #2764 

## What is the new behavior?
A Guide

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/main/.github/CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Reminders
- [-] Make sure you have implemented tests following the guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)".
- [-] Make sure you have updated the cookbook with examples and showcases (for bug fixes, enhancements & new components).

### Review  
- [-] Determine if your changes are a fix, feature or breaking-change, and add the matching label to your PR. If it is tooling, dependency updates or similar, add ignore-for-release.
- [x] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review).
- [x] Request that the changes are code-reviewed 
- [-] Request that the changes are [UX reviewed](https://github.com/kirbydesign/designsystem/blob/main/.github/CONTRIBUTING.md/#ux-review) (only necessary if your PR introduces visual changes)

When the pull request has been approved it will be merged to `develop` by Team Kirby.

### CODE REVIEW

File: libs/designsystem/secondary-entry.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        ├── index.ts
+        ├── ng-packagr.json
+        ├── src
+            ├── index.ts
+            ├── public_api.ts
+            ├── */source files
+```
+
+### 1.a `ng-package.json`
+The `ng-package.json` file is the "hook" which tells `ng-packagr` that there is an entry wihich must be included.
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/secondary-entry.md]:
```suggestion
The `ng-package.json` file is the "hook" which tells `ng-packagr` that there is an entry which must be included.
```

File: libs/designsystem/secondary-entry.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+The file must include an empthy object `{}` and nothing else.
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/secondary-entry.md]:
```suggestion
The file must include an empty object `{}` and nothing else.
```

File: libs/designsystem/secondary-entry.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
+# Move to secondary entry
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/secondary-entry.md:1]: One question that popped up when reading this was "How do I identify a candidate for a secondary entry"? 
E.g. how do I know if a component, service, helper etc. is something that should be moved - would be good to explain that in short here.

File: libs/designsystem/secondary-entry.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+
+## 1. Prepare a secondary entry (SE)
+[ng-pakcagr](https://github.com/ng-packagr/ng-packagr) is a library which Angular uses to pack its code. When we want to make an entry visible for `ng-packagr`, we need the following: 
+
+```
+├── designsystem
+    ├── my-component
+        ├── index.ts
+        ├── ng-packagr.json
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/secondary-entry.md]:
```suggestion
        ├── ng-package.json
```

File: libs/designsystem/secondary-entry.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+``` Typescript
+export * from './public_api.ts';
+```
+
+Index in the root of the SE folder (`my-component`);
+``` Typescript
+export * from './src/index.ts';
+```
+
+## 2. Move the sourcefolder to SE
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/secondary-entry.md]: Consider moving this step up right before step 1a. When following along the guide, I had no files to export from my public_api.ts as described in the current step 1.b.

File: libs/designsystem/secondary-entry.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+It is very important that you DO NOT move the files using drag n' drop! You must use the [`git mv`](https://git-scm.com/docs/git-mv) command; Otherwise git will think that the files was deleted in one place and created in another, which will cause all git history on those files to be lost.
+
+```sh
+git mv libs/designsystem/src/lib/component/<a-component>/* libs/designsystem/my-component/src
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/secondary-entry.md]:
```suggestion
git mv libs/designsystem/src/lib/components/<a-component>/* libs/designsystem/my-component/src
```

File: libs/designsystem/secondary-entry.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+```
+Please note;
+1.  The asterisk (*). This ensures that all the content of the folder is moved. 
+2. The destinationfolder is the `src/` folder, not the SE root folder.
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/secondary-entry.md]: Nice gotchas 🦾

File: libs/designsystem/secondary-entry.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+### 1.a `ng-package.json`
+The `ng-package.json` file is the "hook" which tells `ng-packagr` that there is an entry wihich must be included.
+The file must include an empthy object `{}` and nothing else.
+
+### 1.b `public_api.ts`
+The `public_api.ts` files must export all the source files: 
+
+``` Typescript
+export * from './my.component.ts';
+export * from './my.module.ts';
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/secondary-entry.md]:
```suggestion
export * from './my.component';
export * from './my.module';
```

File: libs/designsystem/secondary-entry.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+```
+
+### 1.c `index.ts`
+As you might have noticed, the filestructure depicted above has two `index.ts` files;
+
+Index in the src file;
+``` Typescript
+export * from './public_api.ts';
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/secondary-entry.md]:
```suggestion
export * from './public_api';
```

File: libs/designsystem/secondary-entry.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+```
+
+Index in the root of the SE folder (`my-component`);
+``` Typescript
+export * from './src/index.ts';
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/secondary-entry.md]:
```suggestion
export * from './src/index';
```

File: libs/designsystem/secondary-entry.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+    "paths": {
+        "@kirbydesign/designsystem/my-component": ["libs/designsystem/my-component/index.ts"],
+    ...
+    }  
+...  
+}
+```
+
+## 4. Change the export
+1. Remove the export from the `src/lib/*` subfolders `index.ts`.
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/secondary-entry.md]: Can we specify here that this is for instance `src/lib/components/index.ts` when moving a component?

File: libs/designsystem/secondary-entry.md (hunk 3 above)
Comment (RasmusKjeldgaard) [reply libs/designsystem/secondary-entry.md:1]: One pointer could be that it should only have dependencies that are already secondary entries.

File: libs/designsystem/secondary-entry.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+@NgModule({
+    imports: [MyModule]
+}) export class AppComponent {}
+```
+In `app.component.html`
+```html
+<my-component></my-component>
+```
+
+Now do the same, but import `kirbyModule`.
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/secondary-entry.md:207]: Is this not what I just did 7.b?

**Review by RasmusKjeldgaard — DISMISSED** [review]
Solid work!

File: libs/designsystem/secondary-entry.md (hunk 5 above)
Comment (RasmusKjeldgaard) [reply libs/designsystem/secondary-entry.md]: So it goes something like: 
1. Make primary entry folder with a src folder in it
2. move files
3. explain and add the rest of the files

File: libs/designsystem/secondary-entry.md [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+Once you have changed all the imports, run the `npm run build` command again.
+The library should now compile.
+
+## 7. Verify the entry
+To verify the move is successfull, do the following;
+1. Run `npm run start` -> navigate to the cookbook page for the component and check that the component looks correct. Also check that there is no errors in the console.
+2. Run `npm run start:flows` -> move around in the app and verify there are no errors in the console.
+
+### 7.a Verify in sperate project
```
Comment (mark-drastrup) [inline libs/designsystem/secondary-entry.md]:
```suggestion
### 7.a Verify in separate project
```

File: libs/designsystem/secondary-entry.md [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+Run `npm run publish` -> install the tarball in a freshly installed angular project.
+
+Import the component and add it to the `app-component.html`
+
+In `app.module.ts`
+```Typescript
+import { MyComponent } from '@kirbydesign/designsystem/my-component';
+
+@NgModule({
+    imports: [MyModule]
```
Comment (mark-drastrup) [inline libs/designsystem/secondary-entry.md:165]:
```suggestion
    imports: [MyComponent]
```

File: libs/designsystem/secondary-entry.md [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+
+### 7.c Verify with primary entry import
+To prevent breaking changes, it must still be possible to import the component through the primary entry:
+
+In `app.module.ts`
+```Typescript
+import { MyComponent } from '@kirbydesign/designsystem';
+
+@NgModule({
+    imports: [MyModule]
```
Comment (mark-drastrup) [inline libs/designsystem/secondary-entry.md:197]:
```suggestion
    imports: [MyComponent]
```

**Review by mark-drastrup — CHANGES_REQUESTED** [review]
Really great write-up! I only have some minor text corrections :)

**Review by mark-drastrup — APPROVED** [review]

**Pull request merged**